File: tests/wasm/test_ewma_volatility.js
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { assertArrayClose, isNaN } from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;

function geometricSeries(length, start, ratio) {
    const out = new Float64Array(length);
    let value = start;
    for (let i = 0; i < length; i++) {
        out[i] = value;
        value *= ratio;
    }
    return out;
}

function pinePeriod(lambdaValue) {
    return Math.max(1, Math.round(2 / (1 - lambdaValue) - 1));
}

function manualEwmaVolatility(data, lambdaValue) {
    const period = pinePeriod(lambdaValue);
    const alpha = 2 / (period + 1);
    const out = new Float64Array(data.length);
    out.fill(NaN);
    const sqReturns = [];
    let seeded = false;
    let ema = NaN;

    for (let i = 1; i < data.length; i++) {
        const prev = data[i - 1];
        const curr = data[i];
        if (Number.isFinite(prev) && Number.isFinite(curr) && prev > 0 && curr > 0) {
            const ret = Math.log(curr / prev);
            const sq = ret * ret;
            sqReturns.push([i, sq]);
            if (!seeded && sqReturns.length === period) {
                seeded = true;
                let sum = 0;
                for (let j = 0; j < sqReturns.length; j++) {
                    sum += sqReturns[j][1];
                }
                ema = sum / period;
                out[i] = Math.sqrt(Math.max(ema, 0)) * 100;
            } else if (seeded) {
                ema = (1 - alpha) * ema + alpha * sq;
                out[i] = Math.sqrt(Math.max(ema, 0)) * 100;
            }
        } else if (seeded) {
            out[i] = out[i - 1];
        }
    }

    if (seeded) {
        for (let i = 1; i < out.length; i++) {
            if (Number.isNaN(out[i]) && !Number.isNaN(out[i - 1])) {
                out[i] = out[i - 1];
            }
        }
    }

    return out;
}

test.before(async () => {
    const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
    const importPath = process.platform === 'win32'
        ? 'file:///' + wasmPath.replace(/\\/g, '/')
        : wasmPath;
    wasm = await import(importPath);
});

test('ewma_volatility_js constant return level', () => {
    const data = geometricSeries(128, 100, 1.01);
    const result = wasm.ewma_volatility_js(data, 0.94);
    const expected = Math.abs(Math.log(1.01)) * 100;
    const period = pinePeriod(0.94);

    for (let i = 0; i < period; i++) {
        assert.ok(isNaN(result[i]), `expected warmup NaN at ${i}`);
    }
    for (let i = period; i < result.length; i++) {
        assert.ok(Math.abs(result[i] - expected) <= 1e-12, `unexpected value at ${i}`);
    }
});

test('ewma_volatility_js matches manual reference', () => {
    const data = geometricSeries(96, 25, 1.003);
    data[12] = NaN;
    data[13] = 0;
    const expected = manualEwmaVolatility(data, 0.90);
    const result = wasm.ewma_volatility_js(data, 0.90);
    assertArrayClose(result, expected, 1e-12, 'manual reference mismatch');
});

test('ewma_volatility_batch_js first row matches single', () => {
    const data = geometricSeries(128, 100, 1.002);
    const batch = wasm.ewma_volatility_batch_js(data, {
        lambda_range: [0.90, 0.94, 0.02],
    });
    const single = wasm.ewma_volatility_js(data, 0.90);

    assert.strictEqual(batch.rows, 3);
    assert.strictEqual(batch.cols, data.length);
    assert.strictEqual(batch.combos.length, 3);
    assert.ok(Math.abs(batch.combos[0].lambda - 0.90) <= 1e-12);
    assertArrayClose(batch.values.slice(0, data.length), single, 1e-12, 'batch row mismatch');
});

test('ewma_volatility_into pointer path matches safe API', () => {
    const data = geometricSeries(128, 75, 1.004);
    const safe = wasm.ewma_volatility_js(data, 0.94);
    const inPtr = wasm.allocate_f64_array(data.length);
    const outPtr = wasm.ewma_volatility_alloc(data.length);

    try {
        wasm.write_f64_array(inPtr, data);
        wasm.ewma_volatility_into(inPtr, outPtr, data.length, 0.94);
        const pointed = wasm.read_f64_array(outPtr, data.length);
        assertArrayClose(pointed, safe, 1e-12, 'pointer path mismatch');
    } finally {
        wasm.ewma_volatility_free(outPtr, data.length);
        wasm.deallocate_f64_array(inPtr);
    }
});

test('ewma_volatility_batch_into returns row count', () => {
    const data = geometricSeries(64, 50, 1.002);
    const inPtr = wasm.allocate_f64_array(data.length);
    const rows = 3;
    const outPtr = wasm.ewma_volatility_alloc(rows * data.length);

    try {
        wasm.write_f64_array(inPtr, data);
        const returnedRows = wasm.ewma_volatility_batch_into(
            inPtr,
            outPtr,
            data.length,
            0.90,
            0.94,
            0.02,
        );
        assert.strictEqual(returnedRows, rows);
    } finally {
        wasm.ewma_volatility_free(outPtr, rows * data.length);
        wasm.deallocate_f64_array(inPtr);
    }
});

test('ewma_volatility_js rejects invalid lambda', () => {
    const data = geometricSeries(32, 10, 1.01);
    assert.throws(() => {
        wasm.ewma_volatility_js(data, 1.0);
    }, /Invalid lambda/);
});

test('ewma_volatility_batch_js rejects invalid config', () => {
    const data = geometricSeries(32, 10, 1.01);
    assert.throws(() => {
        wasm.ewma_volatility_batch_js(data, { lambda_range: [0.94, 0.95] });
    }, /Invalid config/);
});

File: tests/python/test_ewma_volatility.py
import math

import numpy as np
import pytest

try:
    import vector_ta as mp
except ImportError:
    pytest.skip(
        "Python module not built. Run 'maturin develop --features python' first",
        allow_module_level=True,
    )


def geometric_series(length: int, start: float, ratio: float) -> np.ndarray:
    out = np.empty(length, dtype=np.float64)
    value = start
    for i in range(length):
        out[i] = value
        value *= ratio
    return out


def pine_period(lambda_: float) -> int:
    return max(1, int(round(2.0 / (1.0 - lambda_) - 1.0)))


def manual_ewma_volatility(data: np.ndarray, lambda_: float) -> np.ndarray:
    period = pine_period(lambda_)
    alpha = 2.0 / (period + 1.0)
    out = np.full(data.shape[0], np.nan, dtype=np.float64)
    sq_returns = []
    seed_idx = None

    for i in range(1, data.shape[0]):
        prev = data[i - 1]
        curr = data[i]
        if np.isfinite(prev) and np.isfinite(curr) and prev > 0.0 and curr > 0.0:
            ret = math.log(curr / prev)
            sq = ret * ret
            sq_returns.append((i, sq))
            if len(sq_returns) == period and seed_idx is None:
                seed_idx = i
                ema = sum(v for _, v in sq_returns[:period]) / period
                out[i] = math.sqrt(max(ema, 0.0)) * 100.0
            elif seed_idx is not None:
                ema = (1.0 - alpha) * ema + alpha * sq
                out[i] = math.sqrt(max(ema, 0.0)) * 100.0
        elif seed_idx is not None:
            out[i] = out[i - 1]

    if seed_idx is not None:
        for i in range(seed_idx + 1, data.shape[0]):
            if np.isnan(out[i]):
                out[i] = out[i - 1]
    return out


class TestEwmaVolatility:
    def test_constant_return_series_matches_expected_level(self):
        data = geometric_series(128, 100.0, 1.01)
        result = mp.ewma_volatility(data, lambda_=0.94)
        expected = abs(math.log(1.01)) * 100.0
        period = pine_period(0.94)

        assert np.all(np.isnan(result[:period]))
        np.testing.assert_allclose(
            result[period:],
            np.full(data.shape[0] - period, expected),
            rtol=0.0,
            atol=1e-12,
        )

    def test_manual_reference_matches_binding(self):
        data = geometric_series(96, 25.0, 1.003)
        data[12] = np.nan
        data[13] = 0.0
        expected = manual_ewma_volatility(data, 0.90)
        result = mp.ewma_volatility(data, lambda_=0.90)
        np.testing.assert_allclose(result, expected, rtol=0.0, atol=1e-12, equal_nan=True)

    def test_stream_matches_batch(self):
        data = geometric_series(96, 50.0, 1.005)
        batch = mp.ewma_volatility(data, lambda_=0.90)
        streamed = []
        stream = mp.EwmaVolatilityStream(0.90)
        for value in data:
            out = stream.update(float(value))
            streamed.append(np.nan if out is None else out)
        np.testing.assert_allclose(
            np.array(streamed),
            batch,
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )

    def test_batch_metadata_and_first_row_match_single(self):
        data = geometric_series(128, 100.0, 1.002)
        result = mp.ewma_volatility_batch(data, lambda_range=(0.90, 0.94, 0.02))
        assert result["rows"] == 3
        assert result["cols"] == data.shape[0]
        np.testing.assert_allclose(result["lambdas"], np.array([0.90, 0.92, 0.94]))

        single = mp.ewma_volatility(data, lambda_=0.90)
        np.testing.assert_allclose(
            result["values"][0],
            single,
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )

    def test_invalid_lambda(self):
        data = geometric_series(32, 10.0, 1.01)
        with pytest.raises(ValueError, match="Invalid lambda"):
            mp.ewma_volatility(data, lambda_=1.0)

    def test_kernel_parity(self):
        data = geometric_series(96, 30.0, 1.004)
        auto = mp.ewma_volatility(data, lambda_=0.94)
        scalar = mp.ewma_volatility(data, lambda_=0.94, kernel="scalar")
        np.testing.assert_allclose(auto, scalar, rtol=0.0, atol=1e-12, equal_nan=True)
